pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	registerapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha1"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// RegisterCallbackFn is the type of the callback function that handlers will provide
+type RegisterCallbackFn func(pluginName string, versions []string, socketPath string) error
+
+// Watcher is the plugin watcher
+type Watcher struct {
+	path     string
+	handlers map[string]RegisterCallbackFn
+	stopCh   chan interface{}
+	fs       utilfs.Filesystem
+	watcher  *fsnotify.Watcher
+	wg       sync.WaitGroup
+	mutex    sync.Mutex
+}
+
+// NewWatcher provides a new watcher
+func NewWatcher(sockDir string) Watcher {
+	return Watcher{
+		path:     sockDir,
+		handlers: make(map[string]RegisterCallbackFn),
+		fs:       &utilfs.DefaultFs{},
+	}
+}
+
+// AddHandler registers a callback to be invoked for a particular type of plugin
+func (w *Watcher) AddHandler(handlerType string, handlerCbkFn RegisterCallbackFn) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	w.handlers[handlerType] = handlerCbkFn
+}
+
+// Creates the plugin directory, if it doesn't already exist.
+func (w *Watcher) createPluginDir() error {
+	glog.Infof("Ensuring Plugin directory at %s ", w.path)
+	err := w.fs.MkdirAll(w.path, 0755)
+	if err != nil {
+		return fmt.Errorf("error (re-)creating driver directory: %s", err)
+	}
+	return nil
+}
+
+// Walks through the plugin directory to discover any existing plugin sockets.
+func (w *Watcher) traversePluginDir() error {
+	files, err := w.fs.ReadDir(w.path)
+	if err != nil {
+		return fmt.Errorf("error reading the plugin directory: %v", err)
+	}
+	for _, f := range files {
+		// Currently only supports flat fs namespace under the plugin directory.
+		// TODO: adds support for hierarchical fs namespace.
+		if !f.IsDir() && filepath.Base(f.Name())[0] != '.' {
+			w.registerPlugin(path.Join(w.path, f.Name()))
```

Comment:
Why do you think this will involve significant latency?

Reply:
This call is blocking with a timeout of 10s which means the worst case is n * 10s (n number of plugins in the directory).

Reply:
i would prefer the first one, making call goroutine at L#86. If we pass create events, would not they all be processed serially at L#182? 
Or another option could be pass create events and then call `registerPlugin` in a goroutine at L#182. 
I still prefer first one because of simplicity. Thoughts?

Reply:
> i would prefer the first one, making call goroutine at L#86. If we pass create events, would not they all be processed serially at L#182?

Isn't that already the case for all other plugins?

Reply:
but that is not related to kubelet start time. Kubelet start time will be impacted by only those plugins, sockets of which are already there and if we register them serially. In the updated patch, `registerPlugin` is being invoked in a goroutine. PTAL

Reply:
This makes no sense, what the original discussion was about is doing registration outside the main thread.

Whether we choose to do registration of multiple plugins serially or in parallel should not impact the Kubelet start time.

I don't have a preference for the option you choose. I do however think that registration of multiple plugin should follow the same pattern at startup or during runtime.

- Option 1 (sending an event in the event channel) has the nice property of having a single point of entry for both step of the lifecycle thus making refactoring easier.
- Option 2 (have registration of multiple plugins in parallel ) also seems like a sensible approach. 

Additionally it seems like you could even combine both approach.